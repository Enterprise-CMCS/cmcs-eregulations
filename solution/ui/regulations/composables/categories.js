import { ref, watchEffect } from "vue";

import useFetch from "./fetch";

import { getExternalCategories, getInternalCategories } from "utilities/api";

export default function useCategories({ apiUrl, isAuthenticated = false }) {
    const combinedCategories = ref({
        loading: true,
        error: null,
        data: [],
    });

    const externalCategories = useFetch({
        method: getExternalCategories,
        apiUrl,
    });

    const internalCategories = useFetch({
        method: getInternalCategories,
        apiUrl,
        needsAuthentication: true,
        isAuthenticated,
    });

    // watchEffect: super watch
    // https://vuejs.org/guide/essentials/watchers.html#watcheffect
    watchEffect(() => {
        combinedCategories.value.loading =
            externalCategories.value.loading ||
            internalCategories.value.loading;

        combinedCategories.value.error =
            externalCategories.value.error || internalCategories.value.error;

        if (!combinedCategories.value.loading) {
            const externalCats = externalCategories.value.data.map(
                (cat, i) => ({
                    ...cat,
                    categoryType: "categories",
                    catIndex: i,
                })
            );

            const internalCats = internalCategories.value.data.map(
                (cat, i) => ({
                    ...cat,
                    categoryType: "intcategories",
                    catIndex: i,
                })
            );

            combinedCategories.value.data = [...externalCats, ...internalCats];
        }
    });

    return combinedCategories;
}
