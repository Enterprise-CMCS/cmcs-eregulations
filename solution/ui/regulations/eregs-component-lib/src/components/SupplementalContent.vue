<template>
    <div>
        <h1 id="subpart-resources-heading">
            {{ activePart }} Resources
        </h1>
        <h2>Documents</h2>
        <slot name="login-banner" />
        <slot name="public-label" />
        <div class="supplemental-content-container">
            <supplemental-content-category
                v-for="category in categories"
                :key="category.name"
                :name="category.name"
                :subcategory="false"
                :description="category.description"
                :supplemental_content="category.supplemental_content"
                :subcategories="category.subcategories"
                :is-fetching="isFetching"
                :is-fr-link-category="category.is_fr_link_category"
                :show-if-empty="category.show_if_empty"
            />
            <simple-spinner v-if="isFetching" />
        </div>
    </div>
    <slot name="authed-documents" />
    <div class="view-all__container">
        <a
            v-if="selectedPart && subparts.length === 1"
            class="show-subpart-resources"
            data-testid="view-all-subpart-resources"
            @click="clearSection"
        >
            <span class="bold">
                View All Subpart {{ subparts[0] }} Documents</span>
            ({{ resourceCount }})
        </a>
    </div>
</template>

<script>
import {
    getExternalCategories,
    getSupplementalContent,
    getSubpartTOC,
} from "utilities/api";
import {
    EventCodes,
    formatResourceCategories,
    getSectionsRecursive,
} from "utilities/utils";

import SimpleSpinner from "./SimpleSpinner.vue";
import SupplementalContentCategory from "./SupplementalContentCategory.vue";

import eventbus from "../eventbus";

function getDefaultCategories() {
    if (!document.getElementById("categories")) return [];

    const rawCategories = JSON.parse(
        document.getElementById("categories").textContent
    );

    return rawCategories.map((c) => {
        const category = JSON.parse(JSON.stringify(c));
        category.subcategories = [];
        return category;
    });
}

const getCategories = async (apiUrl) => {
    let categories = [];

    try {
        categories = await getExternalCategories({
            apiUrl,
        });
    } catch (error) {
        console.error(error);
    }

    return categories;
};

export default {
    components: {
        SupplementalContentCategory,
        SimpleSpinner,
    },

    props: {
        apiUrl: {
            type: String,
            required: false,
            default: "",
        },
        title: {
            type: String,
            required: true,
        },
        part: {
            type: String,
            required: true,
        },
        subparts: {
            type: Array,
            required: false,
            default() {
                return [];
            },
        },
    },

    data() {
        return {
            categories: [],
            isFetching: true,
            selectedPart: undefined,
            resourceCount: 0,
            partDict: {},
        };
    },

    computed: {
        activePart() {
            if (this.selectedPart !== undefined) {
                return this.selectedPart;
            }
            return `Subpart ${this.subparts[0]}`;
        },
    },

    watch: {
        subparts() {
            this.categories = [];
            this.isFetching = true;
            this.fetchContent();
        },
        selectedPart() {
            this.categories = [];
            this.isFetching = true;
            if (this.selectedPart) {
                this.fetchContent(
                    `citations=${this.title}.${this.part}.${
                        this.selectedPart.split(".")[1]
                    }`
                );
            } else {
                this.fetchContent();
            }
        },
    },

    created() {
        let location = "";
        if (window.location.hash) {
            location = this.parseHash(window.location.hash);
        } else {
            this.fetchContent();
        }
        this.fetchContent(location);
        window.addEventListener("hashchange", this.handleHashChange);
    },
    mounted() {
        eventbus.on(EventCodes.SetSection, (args) => {
            this.selectedPart = args.section;
        });
        this.categories = getDefaultCategories();
    },
    beforeUnmount() {
        eventbus.off(EventCodes.SetSection);
    },
    unmounted() {
        window.removeEventListener("hashchange", this.handleHashChange);
    },

    methods: {
        handleHashChange() {
            const location = this.parseHash(window.location.hash);
            this.fetchContent(location);
        },
        parseHash(locationHash) {
            if (window.location.hash === "#main-content") return "";
            if (locationHash.toLowerCase().includes("appendix")) {
                this.selectedPart = undefined;
                return "";
            }

            let section = locationHash.substring(1).replace("-", ".");

            if (section.includes("-")) {
                // eslint-prefer-destructuring, kinda cool
                [section] = section.split("-");
            }

            if (Number.isNaN(section)) {
                return `citations=${this.title}.${this.part}.${section}`;
            }

            this.selectedPart = `ยง ${section}`;
            return `citations=${this.title}.${section}`;
        },
        async fetchContent(location) {
            try {
                // Page size is set to 1000 to attempt to get all resources.
                // Defualt page size of 100 was omitting resources from the right sidebar.
                // Right now no single subpart hits this number so this shouldn't be an issue

                let response = "";
                if (location) {
                    response = await getSupplementalContent({
                        apiUrl: this.apiUrl,
                        builtCitationString: location,
                        pageSize: 1000,
                    });
                }
                await this.getPartDictionary();

                const results = await Promise.all([
                    getCategories(this.apiUrl),
                    getSupplementalContent({
                        apiUrl: this.apiUrl,
                        partDict: this.partDict,
                        pageSize: 1000,
                    }),
                ]);

                const categories = results[0];
                const subpartResponse = results[1];

                this.resourceCount = subpartResponse.count;

                if (response !== "") {
                    this.categories = formatResourceCategories({
                        apiUrl: this.apiUrl,
                        categories: categories.results,
                        resources: response.results,
                    });
                } else {
                    this.categories = formatResourceCategories({
                        apiUrl: this.apiUrl,
                        categories: categories.results,
                        resources: subpartResponse.results,
                    });
                }
            } catch (error) {
                console.error(error);
            } finally {
                this.isFetching = false;
            }
        },
        async getPartDictionary() {
            const sections = await getSubpartTOC({
                apiUrl: this.apiUrl,
                title: this.title,
                part: this.part,
                subPart: this.subparts[0],
            });

            const secList = getSectionsRecursive(sections);

            this.partDict[this.part] = {
                title: this.title,
                subparts: this.subparts,
                sections: secList,
            };
        },
        clearSection() {
            this.selectedPart = undefined;
            this.location = undefined;
            eventbus.emit(EventCodes.ClearSections);
        },
    },
};
</script>
