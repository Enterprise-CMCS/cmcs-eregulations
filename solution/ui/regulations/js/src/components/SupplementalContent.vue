<template>
  <div>
    <a v-if="selectedPart" v-on:click="clearSection" class="show-subpart-resources">
      <span class="bold"> View All Subpart {{subparts[0]}} Resources</span>  ({{resourceCount}})
    </a>
    <h2 v-if="!requested_categories" id="subpart-resources-heading">
      {{ activePart }} Resources
    </h2>
    <div class="supplemental-content-container">
        <supplemental-content-category
            v-for="category in categories"
            :key="category.name"
            :name="category.name"
            :subcategory="false"
            :description="category.description"
            :supplemental_content="category.supplemental_content"
            :sub_categories="category.sub_categories"
            :isFetching="isFetching"
        >
        </supplemental-content-category>
        <simple-spinner v-if="isFetching"></simple-spinner>
    </div>
  </div>
</template>

<script>
import SimpleSpinner from "./SimpleSpinner.vue";
import SupplementalContentCategory from "./SupplementalContentCategory.vue";

import {getSupplementalContentByCategory, v3GetSupplementalContent} from "../../api";
import {EventCodes, formatResourceCategories} from "../../utils";


function getDefaultCategories(){
    if (!document.getElementById("categories")) return[];

    const rawCategories = JSON.parse(
        document.getElementById("categories").textContent
    );
    const rawSubCategories = JSON.parse(
        document.getElementById("sub_categories").textContent
    );

    return rawCategories.map((c) => {
        const category = JSON.parse(JSON.stringify(c));
        category.sub_categories = rawSubCategories.filter(
            (subcategory) => subcategory.parent_id === category.id
        );
        return category;
    });

}

export default {
    components: {
        SupplementalContentCategory,
        SimpleSpinner,
    },

    props: {
        api_url: {
            type: String,
            required: true,
        },
        title: {
            type: String,
            required: true,
        },
        part: {
            type: String,
            required: true,
        },
        sections: {
            type: Array,
            required: false,
            default: [],
        },
        subparts: {
            type: Array,
            required: false,
            default() {
              return []
            },
        },
        getSupplementalContent: {
          type: Function,
          required: false,
          default: v3GetSupplementalContent
        },
        getSupplementalContentByCategory:{
          type: Function,
          required: false,
          default: getSupplementalContentByCategory
        },
        requested_categories:{
          type: String,
          required: false,
          default: ""
        }
    },

    data() {
        return {
            categories: [],
            isFetching: true,
            selectedPart: undefined,
            resourceCount: 0
        };
    },

    computed: {
        params_array: function () {
            return [
                ["sections", this.sections],
                ["subparts", this.subparts],
            ];
        },
        joined_locations: function () {

            const sectionsString = this.sections.reduce((previousValue, section) => `${previousValue}locations=${this.title}.${this.part}.${section}&`, "")

            const subPartString = this.subparts.reduce((previousValue, subpart) => `${previousValue}locations=${this.title}.${this.part}.${subpart}&`, "")

            return sectionsString + subPartString;
        },
        activePart: function(){
          if (this.selectedPart !== undefined) {

            return this.selectedPart

          }
          return `Subpart ${this.subparts[0]}`
        }
    },

    watch: {
        sections() {
            this.categories = [];
            this.isFetching = true;
            this.fetch_content(this.title, this.part);
        },
        subparts() {
            this.categories = [];
            this.isFetching = true;
            this.fetch_content(this.title, this.part);
        },
        selectedPart() {
            this.categories = [];
            this.isFetching = true;
            if (this.selectedPart) {
              this.fetch_content(this.title, this.part, `locations=${this.title}.${this.part}.${this.selectedPart.split('.')[1]}`);
            }
            else{
              this.fetch_content(this.title, this.part);
            }
        },
    },

    created() {
        this.fetch_content(this.title, this.part);
    },

    mounted() {
        this.$root.$on(EventCodes.SetSection, (args) => {
          this.selectedPart = args.section
        })
        this.categories = getDefaultCategories()
    },

    methods: {
        async fetch_content(title, part, location) {
            try {
                if (this.requested_categories.length > 0){
                    // todo convert this to V3 API
                    this.categories = await this.getSupplementalContentByCategory(
                        this.api_url,
                        this.requested_categories.split(",")
                    );
                }
                else {
                    const response = await v3GetSupplementalContent(
                        this.api_url,
                        {locations: location || this.joined_locations}
                    );

                    if (!this.resourceCount) {
                      this.resourceCount = response.length
                    }

                    this.categories = formatResourceCategories(response)
                }

            } catch (error) {
                console.error(error);
            } finally {
                this.isFetching = false;
            }
        },
      clearSection(){
          console.log("Clearing Section")
          this.selectedPart = undefined
      }
    },
};
</script>