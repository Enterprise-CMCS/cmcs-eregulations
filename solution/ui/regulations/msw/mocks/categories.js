const categories = [
    {
        id: 5,
        name: "Subregulatory Guidance",
        description: "SMDLs, SHOs, CIBs, FAQs, SMM",
        order: 400,
        show_if_empty: true,
        abstractcategory_ptr_id: 5,
    },
    {
        id: 11,
        name: "Implementation Resources",
        description:
            "State Technical Assistance, Toolkits, SPA/Waiver Resources",
        order: 500,
        show_if_empty: true,
        abstractcategory_ptr_id: 11,
    },
    {
        id: 19,
        name: "Reports to Congress",
        description:
            "CMS policy analysis and research reports sent to Congress",
        order: 501,
        show_if_empty: true,
        abstractcategory_ptr_id: 19,
    },
    {
        id: 23,
        name: "Oversight Reports",
        description:
            "Reports from HHS Office of Inspector General and U.S. Government Accountability Office",
        order: 600,
        show_if_empty: true,
        abstractcategory_ptr_id: 23,
    },
];
const subCategories = [{}];
export { categories, subCategories };
