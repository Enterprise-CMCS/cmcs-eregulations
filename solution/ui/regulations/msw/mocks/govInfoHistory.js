const history = [
    {
        year: "1996",
        link:
            "https://www.govinfo.gov/content/pkg/CFR-1996-title42-vol3/pdf/CFR-1996-title42-vol3-sec431-10.pdf",
    },
    {
        year: "1997",
        link:
            "https://www.govinfo.gov/content/pkg/CFR-1997-title42-vol3/pdf/CFR-1997-title42-vol3-sec431-10.pdf",
    },
    {
        year: "1998",
        link:
            "https://www.govinfo.gov/content/pkg/CFR-1998-title42-vol3/pdf/CFR-1998-title42-vol3-sec431-10.pdf",
    },
    {
        year: "1999",
        link:
            "https://www.govinfo.gov/content/pkg/CFR-1999-title42-vol3/pdf/CFR-1999-title42-vol3-sec431-10.pdf",
    },
    {
        year: "2000",
        link:
            "https://www.govinfo.gov/content/pkg/CFR-2000-title42-vol3/pdf/CFR-2000-title42-vol3-sec431-10.pdf",
    },
    {
        year: "2001",
        link:
            "https://www.govinfo.gov/content/pkg/CFR-2001-title42-vol3/pdf/CFR-2001-title42-vol3-sec431-10.pdf",
    },
    {
        year: "2002",
        link:
            "https://www.govinfo.gov/content/pkg/CFR-2002-title42-vol3/pdf/CFR-2002-title42-vol3-sec431-10.pdf",
    },
    {
        year: "2003",
        link:
            "https://www.govinfo.gov/content/pkg/CFR-2003-title42-vol3/pdf/CFR-2003-title42-vol3-sec431-10.pdf",
    },
    {
        year: "2004",
        link:
            "https://www.govinfo.gov/content/pkg/CFR-2004-title42-vol3/pdf/CFR-2004-title42-vol3-sec431-10.pdf",
    },
    {
        year: "2005",
        link:
            "https://www.govinfo.gov/content/pkg/CFR-2005-title42-vol3/pdf/CFR-2005-title42-vol3-sec431-10.pdf",
    },
    {
        year: "2006",
        link:
            "https://www.govinfo.gov/content/pkg/CFR-2006-title42-vol4/pdf/CFR-2006-title42-vol4-sec431-10.pdf",
    },
    {
        year: "2007",
        link:
            "https://www.govinfo.gov/content/pkg/CFR-2007-title42-vol4/pdf/CFR-2007-title42-vol4-sec431-10.pdf",
    },
    {
        year: "2008",
        link:
            "https://www.govinfo.gov/content/pkg/CFR-2008-title42-vol4/pdf/CFR-2008-title42-vol4-sec431-10.pdf",
    },
    {
        year: "2009",
        link:
            "https://www.govinfo.gov/content/pkg/CFR-2009-title42-vol4/pdf/CFR-2009-title42-vol4-sec431-10.pdf",
    },
    {
        year: "2010",
        link:
            "https://www.govinfo.gov/content/pkg/CFR-2010-title42-vol4/pdf/CFR-2010-title42-vol4-sec431-10.pdf",
    },
    {
        year: "2011",
        link:
            "https://www.govinfo.gov/content/pkg/CFR-2011-title42-vol4/pdf/CFR-2011-title42-vol4-sec431-10.pdf",
    },
    {
        year: "2012",
        link:
            "https://www.govinfo.gov/content/pkg/CFR-2012-title42-vol4/pdf/CFR-2012-title42-vol4-sec431-10.pdf",
    },
    {
        year: "2013",
        link:
            "https://www.govinfo.gov/content/pkg/CFR-2013-title42-vol4/pdf/CFR-2013-title42-vol4-sec431-10.pdf",
    },
    {
        year: "2014",
        link:
            "https://www.govinfo.gov/content/pkg/CFR-2014-title42-vol4/pdf/CFR-2014-title42-vol4-sec431-10.pdf",
    },
    {
        year: "2015",
        link:
            "https://www.govinfo.gov/content/pkg/CFR-2015-title42-vol4/pdf/CFR-2015-title42-vol4-sec431-10.pdf",
    },
    {
        year: "2016",
        link:
            "https://www.govinfo.gov/content/pkg/CFR-2016-title42-vol4/pdf/CFR-2016-title42-vol4-sec431-10.pdf",
    },
    {
        year: "2017",
        link:
            "https://www.govinfo.gov/content/pkg/CFR-2017-title42-vol4/pdf/CFR-2017-title42-vol4-sec431-10.pdf",
    },
    {
        year: "2018",
        link:
            "https://www.govinfo.gov/content/pkg/CFR-2018-title42-vol4/pdf/CFR-2018-title42-vol4-sec431-10.pdf",
    },
    {
        year: "2019",
        link:
            "https://www.govinfo.gov/content/pkg/CFR-2019-title42-vol4/pdf/CFR-2019-title42-vol4-sec431-10.pdf",
    },
    {
        year: "2020",
        link:
            "https://www.govinfo.gov/content/pkg/CFR-2020-title42-vol4/pdf/CFR-2020-title42-vol4-sec431-10.pdf",
    },
    {
        year: "2021",
        link:
            "https://www.govinfo.gov/content/pkg/CFR-2021-title42-vol4/pdf/CFR-2021-title42-vol4-sec431-10.pdf",
    },
    {
        year: "2022",
        link:
            "https://www.govinfo.gov/content/pkg/CFR-2022-title42-vol4/pdf/CFR-2022-title42-vol4-sec431-10.pdf",
    },
];
export { history };
