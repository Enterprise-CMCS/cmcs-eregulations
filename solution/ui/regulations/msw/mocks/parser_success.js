const titleFourtyTwoSuccess = {
    id: 3,
    start: "2023-06-28T19:49:36",
    end: "2023-06-28T19:52:06",
    title: 42,
    subchapters: "IV-C",
    parts: "400, 457, 460",
    workers: 3,
    totalVersions: 66,
    skippedVersions: 70,
    errors: 0,
};
export { titleFourtyTwoSuccess };
