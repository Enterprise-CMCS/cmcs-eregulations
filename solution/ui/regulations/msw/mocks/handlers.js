import { http, HttpResponse } from "msw";
import { partToc42, partToc45 } from "./part_toc.js";
import { titles } from "./titles.js";
import { subpartResources } from "./resources.js";
import { subpartA } from "./subpartTOC.js";
import { noEndCondition, titleFourtyTwoSuccess } from "./parser_success.js";
import { history } from "./govInfoHistory.js";

const handlers = [
    http.get("*/title/42/parts", () => {
        return HttpResponse.json(partToc42);
    }),
    http.get("*/title/45/parts", () => {
        return HttpResponse.json(partToc45);
    }),
    http.get("*/titles", () => {
        return HttpResponse.json(titles);
    }),
    http.get("*/title/42/part/433/version/latest/subpart/A/toc", () => {
        return HttpResponse.json(subpartA);
    }),
    http.get("*/resources/", () => {
        return HttpResponse.json(subpartResources);
    }),
    http.get("**test/success/ecfr_parser_result/42", () => {
        return HttpResponse.json(titleFourtyTwoSuccess);
    }),
    http.get("**/test/n/a/ecfr_parser_result/42", () => {
        return HttpResponse.json(noEndCondition);
    }),
    http.get("*/title/42/part/431/history/section/10", () => {
        return HttpResponse.json(history);
    }),
];
export default handlers;
