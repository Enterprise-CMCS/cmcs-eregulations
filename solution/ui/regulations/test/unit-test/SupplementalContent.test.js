import { render, screen, fireEvent } from "@testing-library/vue";
import { describe, it, expect, beforeEach, afterEach } from "vitest";
import SupplementalContent from "eregsComponentLib/src/components/SupplementalContent.vue";
import flushPromises from "flush-promises";
import { categories } from "../../msw/mocks/categories";

describe("Supplemental Content", () => {
    beforeEach(() => {
        const docCat = document.createElement("div");
        docCat.id = "categories";
        docCat.textContent = JSON.stringify(categories);
        document.body.appendChild(docCat);

        const url = "http://eregs.com/";

        Object.defineProperty(window, "location", {
            value: new URL(url),
        });
    });

    afterEach(() => {
        const cat = document.getElementById("categories");
        cat.remove();
    });

    it("Populates some content", async () => {
        render(SupplementalContent, {
            props: {
                apiUrl: "http://localhost:8000/",
                title: "42",
                part: "433",
                subparts: ["A"],
            },
        });

        await flushPromises();

        const view = screen.getByText("Subpart A Resources");
        expect(view.id).toBe("subpart-resources-heading");

        const subG = screen.getByText("Subregulatory Guidance");
        expect(subG).toBeTruthy();
    });

    it("Properly shows and hides empty categories", async () => {
        render(SupplementalContent, {
            props: {
                apiUrl: "http://localhost:8000/",
                title: "42",
                part: "433",
                subparts: ["A"],
            },
        });

        await flushPromises();

        const subEmptyButVisible = screen.queryByText("Empty but Visible");
        expect(subEmptyButVisible).toBeTruthy();

        const subEmptyAndHidden = screen.queryByText("Empty and Hidden");
        expect(subEmptyAndHidden).toBeNull();
    });

    it("Clicks a drop down", async () => {
        render(SupplementalContent, {
            props: {
                apiUrl: "http://localhost:8000/",
                title: "42",
                part: "433",
                subparts: ["A"],
            },
        });

        await flushPromises();
        const subG = await screen.getByLabelText(
            "expand Subregulatory Guidance"
        );
        expect(subG).toBeTruthy();

        await fireEvent.click(subG);
        expect(subG.textContent).toStrictEqual("Subregulatory Guidance ");

        const stateMedBtn = await screen.getByText(
            "State Medicaid Director Letter (SMDL)"
        );
        await flushPromises();
        expect(stateMedBtn.classList.contains("visible")).toBe(false);

        await fireEvent.click(stateMedBtn);
        expect(stateMedBtn.classList.contains("visible")).toBe(true);
    });

    it("Navigates to a section then a subpart", async () => {
        window.location.href += "#433-10";

        render(SupplementalContent, {
            props: {
                apiUrl: "http://localhost:8000/",
                title: "42",
                part: "433",
                subparts: ["A"],
            },
        });

        expect(window.location.hash).toEqual("#433-10");

        await flushPromises();
        const heading = screen.getByText("ยง 433.10 Resources");
        expect(heading.id).toBe("subpart-resources-heading");

        let viewAllSubpartRes = screen.getByTestId(
            "view-all-subpart-resources"
        );
        expect(viewAllSubpartRes.textContent).toBe(
            " View All Subpart A Resources (136) "
        );

        const subG = screen.getByLabelText("expand Subregulatory Guidance");
        await fireEvent.click(subG);

        const stateHealth = screen.getByLabelText(
            "expand State Health Official (SHO) Letter"
        );
        await fireEvent.click(stateHealth);
        await fireEvent.click(viewAllSubpartRes);

        await flushPromises();

        viewAllSubpartRes = screen.queryByTestId("view-all-subpart-resources");
        expect(viewAllSubpartRes).toBeFalsy();

        expect(heading.textContent).toBe("Subpart A Resources");

        const relatedStatues = screen.getByLabelText("expand Related Statutes");
        expect(relatedStatues).toBeTruthy();

        await fireEvent.click(relatedStatues);
    });

    it("Checks to see if the snap shot matches", async () => {
        const wrapper = render(SupplementalContent, {
            props: {
                apiUrl: "http://localhost:8000/",
                title: "42",
                part: "433",
                subparts: ["A"],
            },
        });

        await flushPromises();

        expect(wrapper).toMatchSnapshot();
    });
});
