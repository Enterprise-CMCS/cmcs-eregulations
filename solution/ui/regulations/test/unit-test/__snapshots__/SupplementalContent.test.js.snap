// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Supplemental Content > Checks to see if the snap shot matches 1`] = `
{
  "baseElement": <body>
    <div
      id="categories"
    >
      [{"id":5,"name":"Subregulatory Guidance","description":"SMDLs, SHOs, CIBs, FAQs, SMM","order":400,"show_if_empty":true,"abstractcategory_ptr_id":5},{"id":11,"name":"Implementation Resources","description":"State Technical Assistance, Toolkits, SPA/Waiver Resources","order":500,"show_if_empty":true,"abstractcategory_ptr_id":11},{"id":19,"name":"Reports to Congress","description":"CMS policy analysis and research reports sent to Congress","order":501,"show_if_empty":true,"abstractcategory_ptr_id":19},{"id":23,"name":"Oversight Reports","description":"Reports from HHS Office of Inspector General and U.S. Government Accountability Office","order":600,"show_if_empty":true,"abstractcategory_ptr_id":23}]
    </div>
    <div
      id="sub_categories"
    >
      [{},{}]
    </div>
    <div
      id="categories"
    >
      [{"id":5,"name":"Subregulatory Guidance","description":"SMDLs, SHOs, CIBs, FAQs, SMM","order":400,"show_if_empty":true,"abstractcategory_ptr_id":5},{"id":11,"name":"Implementation Resources","description":"State Technical Assistance, Toolkits, SPA/Waiver Resources","order":500,"show_if_empty":true,"abstractcategory_ptr_id":11},{"id":19,"name":"Reports to Congress","description":"CMS policy analysis and research reports sent to Congress","order":501,"show_if_empty":true,"abstractcategory_ptr_id":19},{"id":23,"name":"Oversight Reports","description":"Reports from HHS Office of Inspector General and U.S. Government Accountability Office","order":600,"show_if_empty":true,"abstractcategory_ptr_id":23}]
    </div>
    <div
      id="sub_categories"
    >
      [{},{}]
    </div>
    <div>
      <div>
        <!---->
        <h2
          id="subpart-resources-heading"
        >
           Subpart A Resources 
        </h2>
        <!---->
        <div
          class="supplemental-content-container"
        >
          <div
            class="supplemental-content-category"
          >
            <div
              class="category"
            >
              <div
                class="category-title childless collapsible-title"
              >
                 Subregulatory Guidance 
              </div>
              <span
                class="childless category-description"
              >
                None
              </span>
              <div
                class="category-content invisible display-none"
                data-test="Subregulatory Guidance"
                style="transition: 0.5s; overflow: hidden;"
              />
            </div>
          </div>
          <div
            class="supplemental-content-category"
          >
            <div
              class="category"
            >
              <div
                class="category-title childless collapsible-title"
              >
                 Implementation Resources 
              </div>
              <span
                class="childless category-description"
              >
                None
              </span>
              <div
                class="category-content invisible display-none"
                data-test="Implementation Resources"
                style="transition: 0.5s; overflow: hidden;"
              />
            </div>
          </div>
          <div
            class="supplemental-content-category"
          >
            <div
              class="category"
            >
              <div
                class="category-title childless collapsible-title"
              >
                 Reports to Congress 
              </div>
              <span
                class="childless category-description"
              >
                None
              </span>
              <div
                class="category-content invisible display-none"
                data-test="Reports to Congress"
                style="transition: 0.5s; overflow: hidden;"
              />
            </div>
          </div>
          <div
            class="supplemental-content-category"
          >
            <div
              class="category"
            >
              <div
                class="category-title childless collapsible-title"
              >
                 Oversight Reports 
              </div>
              <span
                class="childless category-description"
              >
                None
              </span>
              <div
                class="category-content invisible display-none"
                data-test="Oversight Reports"
                style="transition: 0.5s; overflow: hidden;"
              />
            </div>
          </div>
          <!---->
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <!---->
      <h2
        id="subpart-resources-heading"
      >
         Subpart A Resources 
      </h2>
      <!---->
      <div
        class="supplemental-content-container"
      >
        <div
          class="supplemental-content-category"
        >
          <div
            class="category"
          >
            <div
              class="category-title childless collapsible-title"
            >
               Subregulatory Guidance 
            </div>
            <span
              class="childless category-description"
            >
              None
            </span>
            <div
              class="category-content invisible display-none"
              data-test="Subregulatory Guidance"
              style="transition: 0.5s; overflow: hidden;"
            />
          </div>
        </div>
        <div
          class="supplemental-content-category"
        >
          <div
            class="category"
          >
            <div
              class="category-title childless collapsible-title"
            >
               Implementation Resources 
            </div>
            <span
              class="childless category-description"
            >
              None
            </span>
            <div
              class="category-content invisible display-none"
              data-test="Implementation Resources"
              style="transition: 0.5s; overflow: hidden;"
            />
          </div>
        </div>
        <div
          class="supplemental-content-category"
        >
          <div
            class="category"
          >
            <div
              class="category-title childless collapsible-title"
            >
               Reports to Congress 
            </div>
            <span
              class="childless category-description"
            >
              None
            </span>
            <div
              class="category-content invisible display-none"
              data-test="Reports to Congress"
              style="transition: 0.5s; overflow: hidden;"
            />
          </div>
        </div>
        <div
          class="supplemental-content-category"
        >
          <div
            class="category"
          >
            <div
              class="category-title childless collapsible-title"
            >
               Oversight Reports 
            </div>
            <span
              class="childless category-description"
            >
              None
            </span>
            <div
              class="category-content invisible display-none"
              data-test="Oversight Reports"
              style="transition: 0.5s; overflow: hidden;"
            />
          </div>
        </div>
        <!---->
      </div>
    </div>
  </div>,
  "debug": [Function],
  "emitted": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "html": [Function],
  "isUnmounted": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "unmount": [Function],
  "updateProps": [Function],
}
`;
